bento/src/main/java/com/yelp/android/bento/core/ListItemTouchCallback.kt

```diff
@@ -0,0 +1,86 @@
+package com.yelp.android.bento.core
+
+import androidx.recyclerview.widget.ItemTouchHelper
+import androidx.recyclerview.widget.RecyclerView
+
+class ListItemTouchCallback(
+        private val component: ComponentGroup,
+        private val callback: OnItemMovedPositionListener
+) : ItemTouchHelper.Callback() {
+
+    private var dragFrom = -1
+    private var dragTo = -1
+
+    override fun canDropOver(
+            recyclerView: RecyclerView,
+            current: RecyclerView.ViewHolder,
+            target: RecyclerView.ViewHolder
+    ): Boolean {
+        // Only allow reorder if it is within the same component.
```

Comment:
I agree that this implementation is limiting. I guess my questions is, do you have any other suggestions? I suppose we could have some default implementation of this in `Component` that does this check, and override it in `ComponentGroup` if needed. Do you think that would be better?

Reply:
Another thing to remember is that the `RecyclerView`'s reordering only works with `ViewHolder`s. If we were to allow entire components to be reordered, we would have to check that their count is only 1 or some weird things might happen.

Reply:
Another thing I'm gonna throw out is that we may not want to allow everything within one component be reorderable. Consider this layout. It is a ListComponent, where we add another item to the end to view/add photos. We probably don't want to be able to rearrange a photo behind the button.
![image](https://user-images.githubusercontent.com/17460119/60282818-ae79c400-98bc-11e9-92d7-b49684eef99a.png)

Reply:
Yeah I can see how this would get complicated, maybe it should be a follow up. I do think we should support reordering of components within a component group though. We would definitely need to figure out a way to guarantee these are single item components to avoid any messy RecyclerView issues. 

I guess for now just forget about it and maybe open an issue as "nice to have".

Reply:
Whoops, looks like Github lost my comment. TBH, I  think this would be difficult to patch in later. Additionally, the current implementation would not work with the example above, which is the primary reason I am implementing this.

One possible solution is to create a function `canDropItem(fromComponent, fromIndex, toComponent, toIndex)` in each `Component` that allows us to control whether a `Component` can take a given item. 

This solution would cover the case where a component doesn't want to allow an item to be dropped in a certain location (like my example above), but does not consider a `ComponentGroup` with `Component`s with a count > 1. It seems like the only way a `ComponentGroup` could be reorderable is if all of its children are `Component`s with a count == 1. I suppose we could do this check, but it wouldn't be the greatest. Additionally, what would we do when a child has a count > 1? Throw an exception? Not reorder? Just roll with it and hope for the best?

Reply:
Ok, I added a solution for this. The only downside I see if that we now have `findRangedComponentWithIndex` and `findReorderTargetAtIndex` which have very similar implementations, but I don't see a way to combine them.